### PR TITLE
First version of hemera-dynamodb-store

### PR DESCRIPTION
**Pull request by nikkusib**
Please check this updated version of hemera-dynamodb-store

### CODE REVIEW

**Comment by StarpTech** [comment]
Hi @nikkusib could you update the repo according to my review here https://github.com/hemerajs/hemera/pull/181

File: index.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+const Hp = require('hemera-plugin')
+const AWS = require('aws-sdk')
+const DynamoStore = require('./store')
+const StorePattern = require('hemera-store/pattern')
+const Joi = require('joi')
+
+function hemeraDynamoStore(hemera, opts, done) {
+  let topic = 'dynamo-store'
+
+  var dynamo = new AWS.DynamoDB(opts.dynamodb)
```
Comment (StarpTech) [inline index.js]: const

File: index.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    const store = new DynamoStore(db)
+    store.removeById(req, cb)
+  })
+
+  hemera.add({
+    topic,
+    cmd: 'updateById',
+    id: Joi.required(),
+    collection: Joi.string().required(),
+    UpdateExpression: Joi.string().required,
```
Comment (StarpTech) [inline index.js]: Please wrap it in a seperate property like `options`. The reason why is that every primitive value is relevant for pattern matching. A good practice is to wrap options in an object to exclude it.

```js
hemera.add({
  topic: 'math',
  cmd: 'foo',
  collection: 'bar',
  options: {
   ....  
  }
})
```

File: index.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
+'use strict'
+
+const Hp = require('hemera-plugin')
+const AWS = require('aws-sdk')
+const DynamoStore = require('./store')
+const StorePattern = require('hemera-store/pattern')
+const Joi = require('joi')
```
Comment (StarpTech) [inline index.js]: You don't have to include it. You will depend on `hemera-joi` and you can use `const joi = hemera.joi`

File: package.json [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  },
+  "homepage": "https://github.com/hemerajs/hemera-dynamodb-store#readme",
+  "dependencies": {
+    "aws-sdk": "^2.173.0",
+    "hemera-plugin": "^0.1.0",
+    "hemera-store": "^1.1.15"
+  },
+  "devDependencies": {
+    "code": "^5.1.2",
+    "dynamodb-local": "0.0.18",
```
Comment (StarpTech) [inline package.json]: This is no dev dependency

File: package.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
+{
+  "name": "hemera-dynamodb-store",
+  "version": "1.0.0",
```
Comment (StarpTech) [inline package.json]: Always start with 0.0.0

File: index.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  hemera.add({
+    topic,
+    cmd: 'findById',
+    id: Joi.required(),
+    collection: Joi.string().required(),
+    ProjectionExpression: Joi.string(),
+    ExpressionAttributeNames: Joi.object()
+  }, function (req, cb) {
+    const store = new DynamoStore(db)
+    store.findById(req, cb)
```
Comment (StarpTech) [inline index.js]: If you prefer you can also use promise style. It's much better than callback style in this case. E.g https://github.com/hemerajs/hemera-mongo-store/blob/master/index.js#L97

**Comment by StarpTech** [comment]
Hi @nikkusib do you still working on it?

**Comment by nikkusib** [comment]
Hi @StarpTech . I wasn't able to work yesterday, I am working at the moment to fix everything that you have mentioned. One question, I saw that you have created an issue that I don't have travis.yml file included. So I need to create that file with the content of the file that you have provided or I need to add and something else in it. I don't have previous experience with tools like travis so I have to ask :) .

File: package.json (hunk 4 above)
Comment (nikkusib) [reply package.json]: You think on dynamodb-local, right ? So if it is not dev dependency, you think to put in dependencies ?

**Comment by StarpTech** [comment]
Hi @nikkusib this was just an example file. The provided file should serve as an example. Try it otherwise, I will do it.

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "hemera-dynamodb-store",
-  "version": "1.0.0",
+  "version": "0.0.2",
```
Comment (StarpTech) [inline package.json]: You also released it? When not reset 0.0.0

File: package.json [hunk 8]
```diff
@@ -43,7 +44,6 @@
     "eslint-plugin-prettier": "^2.4.0",
     "eslint-plugin-promise": "^3.6.0",
     "eslint-plugin-standard": "^3.0.1",
-    "hemera-joi": "^3.0.0",
     "hemera-testsuite": "^1.4.1",
```
Comment (StarpTech) [inline package.json]: Use latest 2.0.3

File: store.js [hunk 9]
```diff
@@ -36,7 +36,7 @@ class DynamoStore extends Store {
       Item: req.data
     }
     this._driver.put(params, function (err, data) {
-      if (err) cb(err)
+      if (err) return err
```
Comment (StarpTech) [inline store.js]: Why don't you pass the error to the callback?

**Review by StarpTech — APPROVED** [review]

File: package.json (hunk 7 above)
Comment (nikkusib) [reply package.json]: Yes I have published it on npm. What version should I now write to package.json ? 0.0.0 ?

Comment (StarpTech) [reply package.json]: no, when you already published it then we will drive with semver.

File: store.js (hunk 9 above)
Comment (StarpTech) [reply store.js]: ?

File: store.js [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        'id': req.id
+      },
+      ReturnValues: 'ALL_NEW'
+    } 
+    if(req.options) {
+     params = Object.assign(params, req.options)
+    }
+ 
+    this._driver.update(params, function (err, data) {
+      if (err) cb(err)
```
Comment (StarpTech) [inline store.js]: You will call `cb` twice go with this
```js
if(err) {
cb(err)
return
}
cb(null, err)
```

File: store.js [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    const params = {
+      TableName: req.collection,
+      Item: req.data
+    }
+    this._driver.put(params, function (err, data) {
+      if (err){
+        cb(err)
+        return
+      } 
+      cb(null, params.Item)
```
Comment (StarpTech) [inline store.js]: You wanted to change it to `cb(null, data)`

**Review by StarpTech — APPROVED** [review]

**Pull request merged**